internal/controller/runtime/runtime_controller.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtime
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kyma-project/infrastructure-manager/internal/controller/runtime/fsm"
+
+	"k8s.io/client-go/tools/record"
+
+	"github.com/go-logr/logr"
+	imv1 "github.com/kyma-project/infrastructure-manager/api/v1"
+	gardener "github.com/kyma-project/infrastructure-manager/internal/gardener"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// RuntimeReconciler reconciles a Runtime object
+
+type RuntimeReconciler struct {
+	client.Client
+	Scheme        *runtime.Scheme
+	ShootClient   gardener.ShootClient
+	Log           logr.Logger
+	Cfg           fsm.RCCfg
+	EventRecorder record.EventRecorder
+}
+
+//+kubebuilder:rbac:groups=infrastructuremanager.kyma-project.io,resources=runtimes,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infrastructuremanager.kyma-project.io,resources=runtimes/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=infrastructuremanager.kyma-project.io,resources=runtimes/finalizers,verbs=update
+
+var requCounter = 0
+
+func (r *RuntimeReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	r.Log.Info(request.String())
+
+	var runtime imv1.Runtime
+
+	if err := r.Get(ctx, request.NamespacedName, &runtime); err != nil {
+		return ctrl.Result{
+			Requeue: true,
+		}, client.IgnoreNotFound(err)
+	}
+
+	r.Log.Info("Reconciling Runtime", "Name", runtime.Name, "Namespace", runtime.Namespace)
+
+	stateFSM := fsm.NewFsm(
+		r.Log.WithName(fmt.Sprintf("reqID %d", requCounter)),
```

Comment:
Why we don't set runtime as context? I don't see how the `requCounter` will be updated, and how I can correlate it with the particular runtime.